File: GameLibrary/src/cc/lib/zombicide/ZZombie.java
package cc.lib.zombicide;

import cc.lib.game.AGraphics;
import cc.lib.game.GColor;
import cc.lib.game.GDimension;
import cc.lib.game.Justify;
import cc.lib.game.Utils;
import cc.lib.utils.Table;

public final class ZZombie extends ZActor<ZZombieType> {

    static {
        addAllFields(ZZombie.class);
    }

    final ZZombieType type;
    private int imageIdx = -1;
    boolean destroyed = false;

    @Override
    protected int getActionsPerTurn() {
        return type.actionsPerTurn;
    }

    public ZZombie() {
        this(ZZombieType.Walker, -1);
    }

    public ZZombie(ZZombieType type, int zone) {
        super(zone);
        this.type = type;
        onBeginRound();
    }

    private int getIdx() {
        if (imageIdx < 0 || imageIdx >= type.imageOptions.length)
            imageIdx = Utils.rand() % type.imageOptions.length;
        return imageIdx;
    }

    @Override
    public String name() {
        return type.name();
    }

    @Override
    public GDimension drawInfo(AGraphics g, ZGame game, float width, float height) {
        Table info = new Table(getLabel()).setNoBorder();
        info.addRow("Min Hits", type.minDamageToDestroy);
        info.addRow("Actions", type.actionsPerTurn);
        info.addRow("Experience", type.expProvided);
        info.addRow("Ignores Armor", type.ignoresArmor);
        info.addRow("Ranged Priority", type.attackPriority);
        Table outer = new Table().setNoBorder();
        outer.addRow(info, type.description);
        return outer.draw(g);//g.drawString(outer.toString(), 0, 0);
    }

    @Override
    public ZZombieType getType() {
        return type;
    }


    @Override
    public float getScale() {
        return type.getScale();
    }

    @Override
    public int getImageId() {
        return type.imageOptions[getIdx()];
    }

    @Override
    public int getOutlineImageId() {
        return type.imageOutlineOptions[getIdx()];
    }

    @Override
    public GDimension getDimension() {
        if (type.imageDims == null) {
            return GDimension.EMPTY;
        }
        return type.imageDims[getIdx()];
    }

    @Override
    ZCellQuadrant getSpawnQuadrant() {
        switch (type) {
            case Abomination:
            case Wolfbomination:
                return ZCellQuadrant.CENTER;
        }
        return super.getSpawnQuadrant();
    }

    @Override
    long getMoveSpeed() {
        if (type == ZZombieType.Runner) {
            return 500;
        }
        return super.getMoveSpeed();
    }

    @Override
    int getPriority() {
        return type.ordinal();
    }

    public String getDescription() {
        return type.description;
    }

    @Override
    protected boolean performAction(ZActionType action, ZGame game) {
        if (action == ZActionType.MELEE) {
            actionsLeftThisTurn = 0;
            return false; // zombies are done once they attack
        }
        return super.performAction(action, game);
    }

    @Override
    public void draw(AGraphics g) {
        if (isAlive() || isAnimating()) {
            super.draw(g);
            if (actionsLeftThisTurn > 1) {
                g.setColor(GColor.WHITE);
                float oldHgt = g.setTextHeight(10);
                g.drawJustifiedString(getRect().getCenterBottom(), Justify.CENTER, Justify.BOTTOM, String.valueOf(actionsLeftThisTurn));
                g.setTextHeight(oldHgt);
            }
        }
    }

    @Override
    public boolean isAlive() {
        return !destroyed;
    }
}

File: Zombicide/src/cc/lib/zombicide/ZZombie.kt
package cc.lib.zombicide

import cc.lib.game.AGraphics
import cc.lib.game.GColor
import cc.lib.game.GDimension
import cc.lib.game.Justify
import cc.lib.game.Utils
import cc.lib.reflector.Omit
import cc.lib.utils.allMaxOf

open class ZZombie(override val type: ZZombieType = ZZombieType.Walker, val startZone: Int = -1) : ZActor(startZone) {
	companion object {
		init {
			addAllFields(ZZombie::class.java)
		}
	}

	constructor(type: ZZombieType, position: ZActorPosition) : this(type) {
		setPosition(position)
	}

	private var imageIdx = -1

	var destroyed = false
	var frozen = false

	public override val actionsPerTurn: Int
		get() = type.actionsPerTurn

	private val idx: Int
		private get() {
			if (imageIdx < 0 || imageIdx >= type.imageOptions.size) imageIdx =
				Utils.rand() % type.imageOptions.size
			return imageIdx
		}

	override fun name(): String {
        return type.name
    }

    override fun getSpawnQuadrant(): ZCellQuadrant? = when(type) {
        ZZombieType.Abomination,
        ZZombieType.Wolfbomination,
        ZZombieType.BlueTwin,
        ZZombieType.GreenTwin -> ZCellQuadrant.CENTER
        else -> super.getSpawnQuadrant()
    }

    override val scale: Float
        get() = type.scale
    override val imageId: Int
        get() = type.imageOptions[idx]
    override val outlineImageId: Int
        get() = type.imageOutlineOptions[idx]

    override fun getDimension(): GDimension {
        return if (type.imageDims == null) {
            GDimension.EMPTY
        } else type.imageDims[idx]
    }

    override val moveSpeed: Long
        get() = if (type === ZZombieType.Runner) {
	        500
        } else super.moveSpeed

	override val priority: Int
		get() = if (destroyed) -1 else type.ordinal

	fun getDescription(): String {
		return type.description
	}

	override fun isBlockedBy(wallType: ZWallFlag): Boolean = type.isBlockedBy(wallType)

	override suspend fun performAction(action: ZActionType, game: ZGame) {
		when (action) {
			ZActionType.MELEE,
			ZActionType.NOTHING -> actionsLeftThisTurn = 0

			else -> super.performAction(action, game)
		}
	}

	override fun draw(g: AGraphics) {
		if (isAlive || isAnimating) {
			super.draw(g)
            if (actionsLeftThisTurn > 1) {
	            g.pushTextHeight(10f, false)
	            g.color = GColor.TRANSLUSCENT_BLACK
	            val oldWdth = g.setLineWidth(0f)
	            g.drawFilledCircle(getRect().center, .05f)
	            g.color = GColor.WHITE
	            g.drawCircle(getRect().center, 0.05f)
	            g.drawJustifiedString(
		            getRect().center,
		            Justify.CENTER,
		            Justify.CENTER,
		            "$actionsLeftThisTurn"
	            )
	            g.popTextHeight()
	            g.setLineWidth(oldWdth)
            }
        }
    }

	override fun onBeginRound(game: ZGame) {
		super.onBeginRound(game)
		targetZone = null
		if (type == ZZombieType.Ratz) {
			actionsLeftThisTurn = 1 + game.board.getAllZombies().count { it.type == ZZombieType.Ratz }
		}
	}

	@Omit
	private var targetZone: ZZone? = null

	fun getTargetZone(board: ZBoard): ZZone? {
		return targetZone ?: findTargetZone(board)?.also {
			targetZone = it
		}
	}

	override val isAlive: Boolean
		get() = !destroyed

	open fun isEscaping(board: ZBoard): Boolean = false

	protected open fun findTargetZone(board: ZBoard): ZZone? {
		return board.getAllCharacters().filter {
			it.isVisible && board.canSee(occupiedZone, it.occupiedZone)
		}.takeIf {
			it.isNotEmpty()
		}?.map {
			board.getZone(it.occupiedZone)
		}?.allMaxOf {
			it.noiseLevel
		}?.random() ?: board.getMaxNoiseLevelZones().takeIf {
			it.isNotEmpty()
		}?.filter {
			board.isZoneReachable(this, it.zoneIndex)
		}?.randomOrNull()
	}
}